scanner_test.go

```diff
@@ -450,6 +453,67 @@ func TestDecode(t *testing.T) {
 			})
 		}
 	})
+
+	t.Run("wrap errors correctly", func(t *testing.T) {
+
+		t.Run("wrap error from Decoder", func(t *testing.T) {
+			// Use a int parse error as the type example
+			decoder := ss.FuncTagDecoder(func(field ss.Field) (interface{}, error) {
+				return strconv.ParseInt("not-an-int", 10, 0)
+			})
+			err := ss.Decode(&struct {
+				A int `a:""`
+			}{}, decoder)
+
+			var parseErr *strconv.NumError
+			require.True(t, errors.As(err, &parseErr), "error %#v should wrap %T", err, parseErr)
+			require.Equal(t, parseErr.Err, strconv.ErrSyntax)
+		})
+
+		t.Run("wrap error from nested Decoder", func(t *testing.T) {
+			// Use a int parse error as the type example
+			var decoder ss.TagDecoder
+			decoder = ss.FuncTagDecoder(func(field ss.Field) (interface{}, error) {
+				if field.Name == "A" {
+					// Recurse
+					return decoder, nil
+				}
+				return strconv.ParseInt("not-an-int", 10, 0)
+			})
+
+			type Outer struct {
+				A struct{ B int }
+			}
+			err := ss.Decode(&Outer{}, decoder)
+
+			require.Error(t, err)
+			// Sanity check: the outer error _does_ contain the string we don't want to see in the wrapped error
+			require.Contains(t, err.Error(), "error decoding nested field")
+
+			var parseErr *strconv.NumError
+			require.True(t, errors.As(err, &parseErr), "error %#v should wrap %T", err, parseErr)
+			require.Equal(t, parseErr.Err, strconv.ErrSyntax)
+		})
+
+		t.Run("wrap error from slice conversion", func(t *testing.T) {
+			// Use a int parse error as the type example
+			decoder := ss.FuncTagDecoder(func(field ss.Field) (interface{}, error) {
+				return []string{"not-an-int"}, nil
+			})
+			err := ss.Decode(&struct {
+				A []int `a:""`
+			}{}, decoder)
+
+			require.Error(t, err)
+			// Sanity check: the outer error _does_ contain the string we don't want to see in the wrapped error
+			require.Contains(t, err.Error(), "error converting A[0]")
+
+			// In this case, it should just be a wrapped sting error
+			wrapped := errors.Unwrap(err)
+			require.NotSame(t, wrapped, err)
+			require.NotContains(t, wrapped.Error(), "error converting A[0]")
```

Comment:
Here the `errors.Unwrap()` function would not return the same error, and it is not useful to "unwrap" a string error, we can instead just check if the output error contains what we expected of it, e.g.:

```suggestion
			require.Error(t, err)
			// Sanity check: the outer error _does_ contain the string we don't want to see in the wrapped error
			tt.AssertErrContains(t, err, "error", "converting", "A[0]")
```

Otherwise we would be kind of testing the `errors.Unwrap()` function instead of the behavior of our code.

Reply:
I've got updated it to:

```golang
			tt.AssertErrContains(t, err, "error", "converting", "A[0]")

			// In this case, it should just be a wrapped sting error
			wrapped := errors.Unwrap(err)
			tt.AssertNotEqual(t, wrapped, nil)
```

That we we still test that something is wrapped (because without this the test would still pass if the `%s` was used to Errof)